parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations.hs

```diff
@@ -0,0 +1,60 @@
+module Unison.Codebase.SqliteCodebase.Migrations where
+
+import Control.Monad.Reader
+import qualified Data.Map as Map
+import Data.Time.Clock.POSIX (getPOSIXTime)
+import System.Directory (copyFile)
+import System.FilePath ((</>))
+import U.Codebase.Sqlite.Connection (Connection)
+import U.Codebase.Sqlite.DbId (SchemaVersion (..))
+import qualified U.Codebase.Sqlite.Queries as Q
+import Unison.Codebase (CodebasePath)
+import Unison.Codebase.Init.OpenCodebaseError (OpenCodebaseError (OpenCodebaseUnknownSchemaVersion))
+import qualified Unison.Codebase.Init.OpenCodebaseError as Codebase
+import Unison.Codebase.SqliteCodebase.Migrations.MigrateSchema1To2 (migrateSchema1To2)
+import Unison.Codebase.SqliteCodebase.Paths
+import Unison.Codebase.Type (Codebase, LocalOrRemote (..))
+import Unison.Prelude
+import Unison.Symbol (Symbol)
+import Unison.Var (Var)
+import UnliftIO (MonadUnliftIO)
+import qualified UnliftIO
+
+type Migration m a v = Connection -> Codebase m v a -> m ()
+
+-- | The highest schema that this ucm knows how to migrate to.
+currentSchemaVersion :: SchemaVersion
+currentSchemaVersion = fst . head $ Map.toDescList (migrations @IO @Symbol @())
+
+-- | Mapping from schema version to the migration required to get there.
+-- Each migration may only be run on a schema of its immediate predecessor,
+-- E.g. The migration at index 2 must be run on a codebase at version 1.
+migrations :: forall m v a. (MonadUnliftIO m, Var v) => Map SchemaVersion (Migration m a v)
+migrations =
+  Map.fromList
+    [ (2, migrateSchema1To2)
+    ]
+
+-- | Migrates a codebase up to the most recent version known to ucm.
+-- This is a No-op if it's up to date
+-- Returns an error if the schema version is newer than this ucm knows about.
+ensureCodebaseIsUpToDate :: (MonadUnliftIO m, Var v) => LocalOrRemote -> CodebasePath -> Connection -> Codebase m v a -> m (Either Codebase.OpenCodebaseError ())
+ensureCodebaseIsUpToDate localOrRemote root conn codebase = UnliftIO.try do
+  schemaVersion <- runReaderT Q.schemaVersion conn
+  when (schemaVersion > currentSchemaVersion) $ UnliftIO.throwIO $ OpenCodebaseUnknownSchemaVersion (fromIntegral schemaVersion)
+  let migrationsToRun =
+        Map.filterWithKey (\v _ -> v > schemaVersion) migrations
+  when (localOrRemote == Local && (not . null) migrationsToRun) $ backupCodebase root
```

Comment:
One question I have is whether it makes sense to backup in between each migration step, or whether we're okay with just a single backup before the whole thing?

Open to thoughts, right now I have it doing just a single backup.

Reply:
One backup per number-of-boots-below-latest-migration-revision seems fine to me.

Reply:
Are you asking me to change it or are just saying you're fine with either?